core/vm/eof.go

```diff
@@ -0,0 +1,384 @@
+			return fmt.Errorf("%w for section %d: have %d", ErrTooManyInputs, i, typ.Input)
+		}
+		if typ.Output > maxOutputItems && typ.Output != nonReturningFunction {
+			return fmt.Errorf("%w for section %d: have %d", ErrTooManyOutputs, i, typ.Output)
+		}
+		if typ.MaxStackHeight > maxStackHeight {
+			return fmt.Errorf("%w for section %d: have %d", ErrTooLargeMaxStackHeight, i, typ.MaxStackHeight)
+		}
+
+		types = append(types, typ)
+	}
+	c.Types = types
+
+	// Parse code sections.
+	idx += typesSize
+	code := make([][]byte, len(codeSizes))
+	for i, size := range codeSizes {
+		if size == 0 {
+			return fmt.Errorf("%w for section %d: size must not be 0", ErrInvalidCodeSize, i)
+		}
+		code[i] = b[idx : idx+size]
+		idx += size
+	}
+	c.Code = code
+
+	// Parse containers if any
+	if len(containerSizes) != 0 {
+		containers := make([][]byte, len(containerSizes))
+		for i, size := range containerSizes {
+			if size == 0 {
+				return fmt.Errorf("container size may not be 0, container#:%d", i)
+			}
+			containers[i] = b[idx : idx+size]
+			idx += size
+		}
+		c.SubContainer = containers
+	}
+
+	// Parse data section.
+	c.Data = b[idx : idx+dataSize]
+
+	return nil
+}
+
+// ValidateCode validates each code section of the container against the EOF v1
+// rule set.
+func (c *Container) ValidateCode(jt *JumpTable) error {
+	for i, code := range c.Code {
+		if err := validateCode(code, i, c.Types, jt, len(c.Data)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parseSection decodes a (kind, size) pair from an EOF header.
+func parseSection(b []byte, idx int) (kind, size int, err error) {
+	if idx+3 >= len(b) {
+		return 0, 0, io.ErrUnexpectedEOF
+	}
+	kind = int(b[idx])
+	size = int(binary.BigEndian.Uint16(b[idx+1:]))
+	return kind, size, nil
+}
+
+// parseSectionList decodes a (kind, len, []codeSize) section list from an EOF
+// header.
+func parseSectionList(b []byte, idx int) (kind int, list []int, err error) {
+	if idx >= len(b) {
+		return 0, nil, io.ErrUnexpectedEOF
+	}
+	kind = int(b[idx])
+	list, err = parseList(b, idx+1)
+	if err != nil {
+		return 0, nil, err
+	}
+	return kind, list, nil
+}
+
+// parseList decodes a list of uint16..
+func parseList(b []byte, idx int) ([]int, error) {
+	if len(b) < idx+2 {
+		return nil, io.ErrUnexpectedEOF
+	}
+	count := binary.BigEndian.Uint16(b[idx:])
+	if len(b) <= idx+2+int(count)*2 {
+		return nil, io.ErrUnexpectedEOF
+	}
+	list := make([]int, count) // list of sizes
+	for i := 0; i < int(count); i++ {
+		list[i] = int(binary.BigEndian.Uint16(b[idx+2+2*i:]))
+	}
+	return list, nil
+}
+
+// parseUint16 parses a 16 bit BigEndian unsigned integer.
+func parseUint16(b []byte) (int, error) {
+	if len(b) < 2 {
+		return 0, io.ErrUnexpectedEOF
+	}
+	return int(binary.BigEndian.Uint16(b)), nil
+}
+
+// parseInt16 parses a 16 bit signed integer.
+func parseInt16(b []byte) int {
+	return int(int16(b[1]) | int16(b[0])<<8)
+}
+
+// max returns the maximum of a and b.
+func max(a, b int) int {
```

Comment:
You don't need to define `max` & `min` because they were added as built-ins in go 1.21 - see https://go.dev/doc/go1.21#language

Reply:
Thanks, didn't know this